### PR TITLE
[#268] add message error handling

### LINKED ISSUE #268: [BUG] Pulsar cloud storage connector can't convert PROTO_NATIVE schema to Parquet schema
**Describe the bug**

The sink connector is not available to read topics with the `PROTO_NATIVE` schema.

The exception will be thrown:

```
19:04:35.029 [pulsar-io-cloud-storage-sink-flush-0] ERROR org.apache.pulsar.io.jcloud.sink.BlobStoreAbstractSink - Caught unexpected exception:    org.apache.avro.SchemaParseException: Cannot parse <null> schema
        at org.apache.avro.Schema.parse(Schema.java:1633) ~[java-instance.jar:?]
        at org.apache.avro.Schema$Parser.parse(Schema.java:1430) ~[java-instance.jar:?]
        at org.apache.avro.Schema$Parser.parse(Schema.java:1418) ~[java-instance.jar:?]
        at org.apache.pulsar.io.jcloud.util.AvroRecordUtil.convertToAvroSchema(AvroRecordUtil.java:103) ~[qg2KQyXZ9-7y9GGaCCq0Hg/:?]
        at org.apache.pulsar.io.jcloud.format.ParquetFormat.initSchema(ParquetFormat.java:63) ~[qg2KQyXZ9-7y9GGaCCq0Hg/:?]
        at org.apache.pulsar.io.jcloud.sink.BlobStoreAbstractSink.unsafeFlush(BlobStoreAbstractSink.java:239) ~[qg2KQyXZ9-7y9GGaCCq0Hg/:?]
        at org.apache.pulsar.io.jcloud.sink.BlobStoreAbstractSink.flush(BlobStoreAbstractSink.java:219) ~[qg2KQyXZ9-7y9GGaCCq0Hg/:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
```

**To Reproduce**

1. Create a topic with `PROTO_NATIVE` schema
2. Create a sink connector to read from the topic
3. The sink connector will fail

**Expected behavior**

The sink connector should read from topics with protobbuf or protobuf_native schema.

**Screenshots**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by freeznet**
Fixes #268 #220

### Motivation

- handle non structural data types for AVRO / Parquet / JSON format
- extract ProtobufNative schema and infer the GenericRecord if the message without schema version
- fail the record when exception happens
- add `skipFailedMessages` flag

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Review by sijie — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -121,6 +122,22 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
       batchTimeMs: 1000
     ```
 
+### Data format types
+
+Cloud Storage Sink Connector provides multiple options of output format, which are JSON, Avro, Bytes, or Parquet. The default format is JSON.
```
Comment (Huanli-Meng) [inline README.md:127]:
```suggestion
Cloud Storage Sink Connector provides multiple output format options, including JSON, Avro, Bytes, or Parquet. The default format is JSON.
```

File: README.md [hunk 2]
```diff
@@ -121,6 +122,22 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
...
+With current implementation, there are some limitations for different formats:
+
+- JSON: only support structured Pulsar schema types, including `JSON`, `PROTOBUF`, `AVRO`, and `PROTOBUF_NATIVE`, will try to convert the data to JSON from `String` and `Bytes` message.
```
Comment (Huanli-Meng) [inline README.md:130]:
```suggestion
- JSON: only support structured Pulsar schema types, including `JSON`, `PROTOBUF`, `AVRO`, and `PROTOBUF_NATIVE`. The connector will try to convert the data with a `String` or `Bytes` format to JSON-format data.
```

File: README.md [hunk 3]
```diff
@@ -121,6 +122,22 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
...
+- Avro / Parquet: only support structured Pulsar schema types, including `JSON`, `PROTOBUF`, `AVRO`, and `PROTOBUF_NATIVE`.
+- Bytes: support all Pulsar schema types.
+
+When the connector receiving a message with a schema type that is not supported by the connector, the connector will `fail` the message by default, if you want to skip the non-supported messages, you can set `skipFailedMessages` to `true`.
```
Comment (Huanli-Meng) [inline README.md:134]:
```suggestion
By default, when the connector receives a message with a non-supported schema type, the connector will `fail` the message. If you want to skip the non-supported messages, you can set `skipFailedMessages` to `true`.
```

File: README.md [hunk 4]
```diff
@@ -121,6 +122,22 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
...
+
+### Dead letter topic
```
Comment (Huanli-Meng) [inline README.md:136]:
```suggestion
### Dead-letter topics
```

File: README.md [hunk 5]
```diff
@@ -121,6 +122,22 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
...
+
+To use dead-letter-topic, you need to set `skipFailedMessages` to `false`, and set `--max-redeliver-count` and `--dead-letter-topic` when submit the connector with `pulsar-admin`. For more info about dead letter topic, please refer to the [Pulsar documentation](https://pulsar.apache.org/docs/en/concepts-messaging/#dead-letter-topic).
```
Comment (Huanli-Meng) [inline README.md:138]:
```suggestion
To use a dead-letter topic, you need to set `skipFailedMessages` to `false`, and set `--max-redeliver-count` and `--dead-letter-topic` when submit the connector with the `pulsar-admin` CLI tool. For more info about dead-letter topics, see the [Pulsar documentation](https://pulsar.apache.org/docs/en/concepts-messaging/#dead-letter-topic).
```

File: README.md [hunk 6]
```diff
@@ -121,6 +122,22 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
...
+When dead-letter-topic is enabled, the connector will send the message to the dead-letter-topic when the message is failed to be sent to the Cloud Storage.
```
Comment (Huanli-Meng) [inline README.md:139]:
```suggestion
If a message fails to be sent to the Cloud Storage and there is a dead-letter topic, the connector will send the message to the dead-letter topic.
```

File: docs/io-cloud-storage-sink.md [hunk 7]
```diff
@@ -215,3 +232,15 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
         The date is generated based on the `partitionerType` parameter in the configuration. And the `${messageSequenceId}` is generated by `FunctionCommon.getSequenceId(message.getMessageId())`.
     - Format suffix: `.parquet`
         This part is generated based on the `formatType` parameter in the configuration.
+    
+## Permissions
+
+### AWS S3 permission policies
+
+The suggested permission policies for AWS S3 are:
```
Comment (Huanli-Meng) [inline docs/io-cloud-storage-sink.md:240]: I am not sure whether we can change this sentence to "It is recommended to configure the following AWS S3 permission policies:"

File: docs/io-cloud-storage-sink.md [hunk 8]
```diff
@@ -215,3 +232,15 @@ Before using the Cloud Storage sink connector, you need to create a configuratio
...
+- `s3:AbortMultipartUpload`
```
Comment (Huanli-Meng) [inline docs/io-cloud-storage-sink.md:241]: can we add some descriptions about those policies?

**Review by Huanli-Meng — COMMENTED** [review]
Leave my comments. You can also update my comments to the sink connector doc.

**Pull request merged**